File: TodoListApp_Android/app/src/main/java/io/viper/android/todolist/todolist/data/source/local/TasksLocalDataSource.java
/*
 * Copyright (c) 2016 lhyz Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.viper.android.todolist.todolist.data.source.local;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.stmt.DeleteBuilder;
import com.j256.ormlite.stmt.PreparedDelete;
import com.j256.ormlite.stmt.PreparedQuery;
import com.j256.ormlite.stmt.QueryBuilder;

import java.sql.SQLException;
import java.util.List;

import io.viper.android.todolist.todolist.data.Task;
import io.viper.android.todolist.todolist.data.source.TasksDataSource;

/**
 * hello,android
 * Created by lhyz on 2016/8/3.
 */
public class TasksLocalDataSource implements TasksDataSource {

    private static TasksLocalDataSource INSTANCE;

    Dao<Task, String> mTasksDao;

    private TasksLocalDataSource(Context context) {
        DatabaseHelper databaseHelper = new DatabaseHelper(context);
        try {
            mTasksDao = databaseHelper.getTasksDao();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public static TasksLocalDataSource getInstance(Context context) {
        if (INSTANCE == null) {
            INSTANCE = new TasksLocalDataSource(context);
        }
        return INSTANCE;
    }

    private @Nullable Task findTaskById(String taskId) throws SQLException {
        QueryBuilder<Task, String> builder = mTasksDao.queryBuilder();
        builder.where().eq("taskId", taskId);
        PreparedQuery<Task> preparedQuery = builder.prepare();
        return mTasksDao.queryForFirst(preparedQuery);
    }

    @Override
    public void getAllTasks(@NonNull LoadTasksCallback callback) {
        try {
            List<Task> results = mTasksDao.queryForAll();
            callback.onTasksLoaded(results);
        } catch (SQLException e) {
            callback.onDataNotAvailable();
            e.printStackTrace();
        }
    }

    @Override
    public void getTask(@NonNull String taskId, @NonNull GetTaskCallback callback) {
        try {
            Task task = findTaskById(taskId);
            if (task == null) {
                callback.onDataNotAvailable();
                return;
            }
            callback.onTaskLoaded(task);
        } catch (SQLException e) {
            callback.onDataNotAvailable();
            e.printStackTrace();
        }
    }

    @Override
    public void saveTask(@NonNull Task task) {
        try {
            mTasksDao.create(task);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void updateTask(@NonNull Task task) {
        try {
            mTasksDao.update(task);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void completeTask(@NonNull String taskId) {
        try {
            Task task = findTaskById(taskId);
            if (task == null) {
                return;
            }
            task.setCompleted(true);
            mTasksDao.update(task);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void activateTask(@NonNull String taskId) {
        try {
            Task task = findTaskById(taskId);
            if (task == null) {
                return;
            }
            task.setCompleted(false);
            mTasksDao.update(task);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void clearCompletedTasks() {
        try {
            DeleteBuilder<Task, String> builder = mTasksDao.deleteBuilder();
            builder.where().eq("completed", true);
            PreparedDelete<Task> preparedDelete = builder.prepare();
            mTasksDao.delete(preparedDelete);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void refreshTasks() {
        try {
            List<Task> tasks = mTasksDao.queryForAll();
            for (Task task : tasks) {
                mTasksDao.refresh(task);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void deleteAllTasks() {
        try {
            List<Task> tasks = mTasksDao.queryForAll();
            mTasksDao.delete(tasks);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void deleteTask(@NonNull String taskId) {
        try {
            DeleteBuilder<Task, String> builder = mTasksDao.deleteBuilder();
            builder.where().eq("taskId", taskId);
            PreparedDelete<Task> preparedDelete = builder.prepare();
            mTasksDao.delete(preparedDelete);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: TodoListApp_Kotlin/app/src/main/java/com/lhyz/android/todolist/data/source/local/TasksLocalDataSource.kt
package com.lhyz.android.todolist.data.source.local

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteException
import com.lhyz.android.todolist.data.Task
import com.lhyz.android.todolist.data.source.TasksDataSource
import com.lhyz.android.todolist.data.source.local.TasksPersistenceContract.TaskEntry
import java.util.*

/**
 * hello,android
 * @author lhyz on 2017/2/5
 */
class TasksLocalDataSource private constructor(context: Context) : TasksDataSource {

    companion object {
        fun getInstance(context: Context): TasksLocalDataSource {
            return TasksLocalDataSource(context)
        }
    }

    private val mDbHelper: TasksDbHelper = TasksDbHelper(context)

    override fun getTasks(callback: TasksDataSource.LoadTasksCallback) {
        val tasks = ArrayList<Task>()
        val db = mDbHelper.writableDatabase
        val projection = arrayOf(
            TaskEntry.COLUMN_NAME_TASK_ID,
            TaskEntry.COLUMN_NAME_TITLE,
            TaskEntry.COLUMN_NAME_DESCRIPTION,
            TaskEntry.COLUMN_NAME_COMPLETED
        )
        val c = db.query(TaskEntry.TABLE_NAME, projection, null, null, null, null, null)
        if (c != null && c.count > 0) {
            while (c.moveToNext()) {
                val itemId = c.getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_TASK_ID))
                val title = c.getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_TITLE))
                val description =
                    c.getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_DESCRIPTION))
                val completed =
                    c.getInt(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_COMPLETED)) == 1
                val task = Task(itemId, title, description, completed)
                tasks.add(task)
            }
        }
        c?.close()
        db.close()

        if (tasks.isEmpty()) {
            callback.onDataNotAvailable()
        } else {
            callback.onTasksLoaded(tasks)
        }
    }

    override fun getTask(taskId: String, callback: TasksDataSource.GetTaskCallback) {
        val db = mDbHelper.readableDatabase

        val projection = arrayOf(
            TaskEntry.COLUMN_NAME_TASK_ID,
            TaskEntry.COLUMN_NAME_TITLE,
            TaskEntry.COLUMN_NAME_DESCRIPTION,
            TaskEntry.COLUMN_NAME_COMPLETED
        )

        val selection = TaskEntry.COLUMN_NAME_TASK_ID + " LIKE ?"
        val selectionArgs = arrayOf(taskId)

        val c = db.query(
            TaskEntry.TABLE_NAME, projection, selection, selectionArgs, null, null, null
        )

        var task: Task? = null

        if (c != null && c.count > 0) {
            c.moveToFirst()
            val itemId = c.getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_TASK_ID))
            val title = c.getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_TITLE))
            val description =
                c.getString(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_DESCRIPTION))
            val completed = c.getInt(c.getColumnIndexOrThrow(TaskEntry.COLUMN_NAME_COMPLETED)) == 1
            task = Task(itemId, title, description, completed)
        }
        c?.close()

        db.close()

        if (task != null) {
            callback.onTaskLoaded(task)
        } else {
            callback.onDataNotAvailable()
        }
    }

    override fun saveTask(task: Task) {
        val db = mDbHelper.writableDatabase

        val values = ContentValues()
        values.put(TaskEntry.COLUMN_NAME_TASK_ID, task.taskId)
        values.put(TaskEntry.COLUMN_NAME_TITLE, task.title)
        values.put(TaskEntry.COLUMN_NAME_DESCRIPTION, task.description)
        values.put(TaskEntry.COLUMN_NAME_COMPLETED, task.completed)

        db.insert(TaskEntry.TABLE_NAME, null, values)

        db.close()
    }

    override fun completeTask(task: Task) {
        val db = mDbHelper.writableDatabase

        val values = ContentValues()
        values.put(TaskEntry.COLUMN_NAME_COMPLETED, true)

        val selection = TaskEntry.COLUMN_NAME_TASK_ID + " LIKE ?"
        val selectionArgs = arrayOf(task.taskId)

        db.update(TaskEntry.TABLE_NAME, values, selection, selectionArgs)

        db.close()
    }

    override fun completeTask(taskId: String) {
        throw UnsupportedOperationException("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun activateTask(task: Task) {
        val db = mDbHelper.writableDatabase

        val values = ContentValues()
        values.put(TaskEntry.COLUMN_NAME_COMPLETED, false)

        val selection = TaskEntry.COLUMN_NAME_TASK_ID + " LIKE ?"
        val selectionArgs = arrayOf(task.taskId)

        db.update(TaskEntry.TABLE_NAME, values, selection, selectionArgs)

        db.close()
    }

    override fun activateTask(taskId: String) {
        throw UnsupportedOperationException("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun clearCompletedTasks() {
        val db = mDbHelper.writableDatabase

        val selection = TaskEntry.COLUMN_NAME_COMPLETED + " LIKE ?"
        val selectionArgs = arrayOf("1")

        db.delete(TaskEntry.TABLE_NAME, selection, selectionArgs)

        db.close()
    }

    override fun refreshTasks() {
        throw UnsupportedOperationException("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun deleteAllTasks() {
        val db = mDbHelper.writableDatabase
        try {
            db.delete(TaskEntry.TABLE_NAME, null, null)
        } catch (e: SQLiteException) {
            e.printStackTrace()
        }
        db.close()
    }

    override fun deleteTask(taskId: String) {
        val db = mDbHelper.writableDatabase

        val selection = TaskEntry.COLUMN_NAME_TASK_ID + " LIKE ?"
        val selectionArgs = arrayOf(taskId)

        db.delete(TaskEntry.TABLE_NAME, selection, selectionArgs)

        db.close()
    }
}